tests/group_functions/group_functions_common.h

```diff
@@ -0,0 +1,305 @@
+    sycl::id<3> sizes =
+        queue.get_device().get_info<sycl::info::device::max_work_item_sizes>();
+#else
+    sycl::id<3> sizes =
+        queue.get_device()
+            .get_info<sycl::info::device::max_work_item_sizes<3>>();
+#endif
+    for (int i = 0; i < Dimensions; ++i) {
+      max_work_item_sizes[i] = sizes.get(i);
+    }
+  }
+  size_t max_work_group_size =
+      queue.get_device().get_info<sycl::info::device::max_work_group_size>();
+
+  // make work-group size as much square/cubic as possible
+  size_t work_group_sizes[Dimensions] = {
+      std::min(max_work_item_sizes[0], max_work_group_size)};
+  if constexpr (Dimensions > 1) {
+    size_t rest_work_group_size = max_work_group_size;
+    for (int cur_D = Dimensions; cur_D > 1; --cur_D) {
+      size_t pref_size = pow(rest_work_group_size, 1. / cur_D) + 1;
+      pref_size = std::min(pref_size, max_work_item_sizes[cur_D - 1]);
+      // in the worst case of prime rest_work_group_size pref_size comes to 1
+      while (rest_work_group_size % pref_size != 0) {
+        --pref_size;
+      }
+      work_group_sizes[cur_D - 1] = pref_size;
+      rest_work_group_size /= pref_size;
+    }
+    work_group_sizes[0] =
+        std::min(max_work_item_sizes[0], rest_work_group_size);
+  }
+
+  sycl::range<Dimensions> work_group_range = get_default_range<Dimensions>();
+  for (int i = 0; i < Dimensions; ++i)
+    work_group_range[i] = work_group_sizes[i];
+
+  return work_group_range;
+}
+
+/**
+ * @brief Provides group size pretty printing
+ * @tparam D Dimension of group instance
+ */
+template <int D>
+std::string work_group_print(const sycl::range<D>& work_group_range) {
+  std::string res("{ " + std::to_string(work_group_range[0]));
+  for (int i = 1; i < D; ++i) res += ", " + std::to_string(work_group_range[i]);
+  res += " }";
+  return res;
+}
+
+/** @brief A user-defined class with several scalar member variables, default
+ *         constructor and some overloaded operators.
+ */
+struct custom_type {
+  int m_int_field{};
+  bool m_bool_field{};
+
+  custom_type() = default;
+
+  custom_type(int value) {
+    m_int_field = value;
+    m_bool_field = value;
+  }
+
+  friend bool operator==(const custom_type& c_t_l, const custom_type& c_t_r) {
+    return c_t_l.m_int_field == c_t_r.m_int_field &&
+           c_t_l.m_bool_field == c_t_r.m_bool_field;
+  }
+
+  operator int() const { return m_int_field; }
+
+  void operator=(int value) {
+    m_int_field = value;
+    m_bool_field = value;
+  }
+};
+
+/**
+ * @brief Provides limit for exact integer number representation by different
+ * types
+ *
+ */
+template <typename T>
+inline constexpr uint64_t exact_max = std::numeric_limits<T>::max();
+
+template <>
+inline constexpr uint64_t exact_max<sycl::half> = 1ull << 11;
+template <>
+inline constexpr uint64_t exact_max<float> = 1ull << 24;
+template <>
+inline constexpr uint64_t exact_max<double> = 1ull << 53;
+
+}  // namespace util
+
+// Cartesian product of type lists
```

Comment:
I have the feeling this should be moved into `util/` part of the tests.

Reply:
After reading this, I do not understand why this is so complex. At the end it should be just reduced to a nested variadic pack expansion.

Reply:
This product can be skipped and removed totally by using `for_all_combinations` everywhere, but it slightly extend the code for invocations (a helper class should be written for every test function, not only for scans).